ui/src/app/views/workspace/component.ts

```diff
@@ -191,10 +195,38 @@ export class WorkspaceComponent implements OnInit {
   }
 
   openCluster(): void {
-    const url = 'https://leonardo.dsde-dev.broadinstitute.org/notebooks/'
+    const leoBaseUrl = 'https://leonardo.dsde-dev.broadinstitute.org';
+    const leoNotebookUrl = leoBaseUrl + '/notebooks/'
         + this.cluster.clusterNamespace + '/'
         + this.cluster.clusterName;
-    window.location.href = url;
+    const leoSetCookieUrl = leoNotebookUrl + '/setCookie';
+
+    const headers = new Headers();
+    headers.append('Authorization', 'Bearer ' + this.signInService.currentAccessToken);
+    this.http.get(leoSetCookieUrl, {
+      headers: headers,
+      withCredentials: true
+    }).subscribe(() => {
+      const notebook = window.open(leoNotebookUrl, '_blank');
+      this.clusterPulled = false;
+      window.addEventListener('message', (e) => {
```

Comment:
Should we potentially add this listener before the call to window.open? (Wondering if there could be a race condition here.)

Also: what happens if you hit launch notebook multiple times... do you get multiple listeners registered?

Reply:
I see no problem adding that before the window.open call. 

Currently I believe you do get multiple listeners registered. I'll add a bool that gets flipped to true the first time.

Reply:
Actually, it does need to be before the call to window.open, since we need the return value of that call for use in the event listener.

Reply:
If you're going to flip a boolean to "true" the first time... you may need to ensure that a consistent target window is used each time. Otherwise newly-opened windows won't have a way to learn the client ID to use. (Alternatively, we need to be sure that the JS client code running in the notebook extension is [setting its cookie](https://github.com/DataBiosphere/leonardo/blob/cfc455a6c3f0ee7571371aa96b334c8b3ab76cda/src/main/resources/jupyter/google_sign_in.js#L69) at a path like `/notebooks` instead of the default `/notebooks/:project/:cluster` ... otherwise tokens will expire at the root path, I believe. This is complicated; happy to discuss offline!)